File: problems/回溯/78.子集.cpp
#include <iostream>
#include <vector>
using namespace std;

void backtrack(vector<vector<int>>& result, vector<int>& path, 
               vector<int>& nums, int start) {
    result.push_back(path);
    for (int i = start; i < nums.size(); i++) {
        path.push_back(nums[i]);
        backtrack(result, path, nums, i + 1);
        path.pop_back();
    }
}

vector<vector<int>> subsets(vector<int>& nums) {
    vector<vector<int>> result;
    vector<int> path;
    backtrack(result, path, nums, 0);
    return result;
}

int main() {
    // 测试用例1
    vector<int> nums1 = {1, 2, 3};
    vector<vector<int>> res1 = subsets(nums1);
    cout << "Test 1: [1,2,3]" << endl;
    for (auto& subset : res1) {
        cout << "[";
        for (int i = 0; i < subset.size(); i++) {
            cout << subset[i];
            if (i < subset.size() - 1) cout << ",";
        }
        cout << "] ";
    }
    cout << endl;
    
    // 测试用例2
    vector<int> nums2 = {0};
    vector<vector<int>> res2 = subsets(nums2);
    cout << "Test 2: [0]" << endl;
    for (auto& subset : res2) {
        cout << "[";
        for (int i = 0; i < subset.size(); i++) {
            cout << subset[i];
            if (i < subset.size() - 1) cout << ",";
        }
        cout << "] ";
    }
    cout << endl;
    
    // 测试用例3
    vector<int> nums3 = {1, 2, 3, 4};
    vector<vector<int>> res3 = subsets(nums3);
    cout << "Test 3: [1,2,3,4] - " << res3.size() << " subsets" << endl;
    
    return 0;
}

File: problems/回溯/78.子集.py
#!/usr/bin/env python3
"""
78. 子集 (Subsets)
题目: 给你一个整数数组 nums ，返回该数组所有可能的子集（幂集）。子集不能包含重复元素。
算法: 回溯
时间复杂度: O(2^n)
空间复杂度: O(n)
"""

def subsets(nums):
    result = []
    path = []
    
    def backtrack(start):
        result.append(path[:])
        for i in range(start, len(nums)):
            path.append(nums[i])
            backtrack(i + 1)
            path.pop()
    
    backtrack(0)
    return result

if __name__ == "__main__":
    # 测试用例1
    print(subsets([1, 2, 3]))
    # 输出: [[], [1], [1,2], [1,2,3], [1,3], [2], [2,3], [3]]
    
    # 测试用例2
    print(subsets([0]))
    # 输出: [[], [0]]
    
    # 测试用例3
    print(subsets([1, 2, 3, 4]))
    # 输出: 16个子集
